apis/experimental/node_classes/types.go

```diff
@@ -0,0 +1,152 @@
+	NodePoolTerminated NodePoolPhase = "Terminated"
+	// ...
+)
+
+// These are objects that a cluster admin (or infrastructure admin, if a
+// differnet role) would create in order to set the policy of what kinds of
+// nodes are available and allowed for the cluster, analogous to
+// PersistentVolumes and PersistentVolumeClaims. They can have generic names
+// like "small," "medium," "gpu," or as arbitrarily complex as is useful
+// ("large-gpu-us-west-ssd"). In a cloud context, it allows infrastructure
+// admins to place restrictions on what resources are allowed to be created,
+// and in an on-premise context, allows accurate modeling of what hardware is
+// actually available for use.
+//
+// These serve as the glue between NodePools and NodeBuilders. In order to keep
+// the NodeClass registry from becoming a combinatorial explosion, we can
+// parameterize attributes that aren't core to the NodeClass' identity. For
+// instance, it might be very important that region/zone is a core part of a
+// NodeClass, and an admin wants their naming scheme to always start with the
+// zone. However, within that zone, the exact disk image you use is far more
+// flexible and inconsequential to the cluster admin, so it's modeled as an
+// overridable attribute. NodePools can then specify Args to fill in the blanks
+// or override defaults, and NodeBuilders will merge the Args into the Params
+// and use the result to populate a node's attributes.
+type NodeClass struct {
+	Name string
+	Params NodeClassParams
+	NodeBuilder api.ObjectReference
+}
+
+type NodeClassParams struct {
+	InstanceType string // n1-standard-1, r2.medium, etc.
+	Disks []Disk
+}
+
+// Open question: is there value in modeling key package versions for a node?
+// Can be represented in an OS-agnostic way a la cloud-init, but would let us
+// have much more control over version of Docker, kubelet, etc., and perform
+// rolling updates (potentially in-place).
+type Disk struct {
+	Bootable bool
+	Size uint64
+	Image string
+}
+
+// This would be attached to control plane components and NodeSpec so that you
+// can trigger upgrades.
+type KubernetesVersion struct {
+	SemanticVersion string
+	Path string // gs://..., s3://...
+}
+
+// NodeBuilders are the concrete cloud-provider definitions. They should have
+// enough configuration to be able to make the service calls necessary to
+// create/destroy/manipulate VMs, etc. The controller that ultimately makes a
+// call to AWS to ask for a new node would be watching the AWSNodeBuilder
+// objects and all NodePool objects that are linked to them in order to be able
+// to reconcile state.
+//
+// These would likely be created by the administrator (or cluster provisioning
+// tool) as a way of linking cloud service accounts and settings with the
+// cluster.
+//
+// There's an open question about how these should link to the Cluster object,
+// especially if we want to support representing N clusters as concurrent
```

Comment:
I think that this API should punt on the N clusters case and focus on managing a single cluster. The registry can be the aggregation layer.

Reply:
SGTM (until we have a good reason to want it).